src/Storage/Hydration/EventHydrator.php

```diff
@@ -0,0 +1,167 @@
+        }
+
+        $class = $this->classResolver->resolve($data['eventType']);
+
+        try {
+            $reflection = new ReflectionClass($class);
+        } catch (ReflectionException $exception) {
+            throw new HydrationFailed(
+                sprintf("Cannot hydrate event, class '%s' could not be found.", $class),
+                0,
+                $exception
+            );
+        }
+
+        $event      = $reflection->newInstanceWithoutConstructor();
+        $properties = $this->getProperties($reflection);
+
+        if (!($event instanceof Event)) {
+            throw new HydrationFailed(
+                sprintf("Cannot hydrate event as the class '%s' is not an instance of event.", $class)
+            );
+
+        }
+
+        foreach ($properties as $property) {
+            $this->setEventProperty($event, $property, $data);
+        }
+
+        return $event;
+    }
+
+
+
+    /**
+     * @return mixed[]
+     */
+    public function dehydrate(Event $event) : array
+    {
+        $data = [];
+
+        $reflection = new ReflectionClass($event);
+        $properties = $this->getProperties($reflection);
+
+        foreach ($properties as $property) {
+            $data[$property->getName()] = $this->getEventPropertyValue($event, $property);
+        }
+
+        $data['eventType'] = $event::eventType();
+
+        return $data;
+    }
+
+
+    /**
+     * @return mixed
+     */
+    private function getEventPropertyValue(Event $event, ReflectionProperty $property)
+    {
+        // First check to see if we have a getter method. If we do, use that to get the value instead.
+        $method = $this->findGetterForProperty($property);
+
+        if ($method) {
+            return $method->invoke($event);
+        }
+
+        $property->setAccessible(true);
+
+        return $property->getValue($event);
+    }
+
+
+    /**
+     * @param mixed[] $data
+     */
+    private function setEventProperty(Event $event, ReflectionProperty $property, array $data) : void
+    {
+        if (!isset($data[$property->getName()])) {
+            return;
+        }
+
+        $property->setAccessible(true);
+        $property->setValue($event, $data[$property->getName()]);
+    }
+
+
+    private function findGetterForProperty(ReflectionProperty $property) : ?ReflectionMethod
+    {
+        $reflection = $property->getDeclaringClass();
+        $methodName = sprintf('get%s', ucfirst($property->getName()));
+        $method     = $this->getMethodOrNull($reflection, $methodName);
+
+        // If we can't find a method explicitly prefixed with 'get' then try finding a property matching the parameter
+        // name exactly.
+        if (!$method) {
+            $method = $this->getMethodOrNull($reflection, $property->getName());
+        }
+
+        // Only use the method if it looks like a getter.
+        if ($method && $method->isPublic() && $method->hasReturnType() && $method->getNumberOfParameters() === 0) {
```

Comment:
`$method->hasReturnType()` - is this strictly necessary? We might be dealing with a getter which does have a return type, but doesn't have one explicitly hinted. Do we want to support that case?

Reply:
I'm not sure. It seemed like a reasonable requirement since this library is `php >=7.2` though having said that it won't be supported if your return type is `mixed`. On the other hand there's a change we end up picking a method that doesn't return anything.

Reply:
Maybe this check should just be for non-void return types where one is defined, but allow unhinted methods to support mixed.

Reply:
Done